source/jormungandr/jormungandr/instance.py

```diff
@@ -549,20 +556,45 @@ def reap_socket(self, ttl):
                     break  # remaining socket are still in "keep alive" state
             except IndexError:
                 break
+                now = time.time()
+        while True:
+            try:
+                socket, t = self._pt_sockets.popleft()
+                if now - t > ttl:
+                    logger.debug("closing one socket for %s", self.name)
+                    socket.setsockopt(zmq.LINGER, 0)
+                    socket.close()
+                else:
+                    self._pt_sockets.appendleft((socket, t))
+                    break  # remaining socket are still in "keep alive" state
+            except IndexError:
+                break
 
     @contextmanager
-    def socket(self, context):
+    def socket(self, context, pt_socket=False):
         socket = None
-        try:
-            socket, _ = self._sockets.pop()
-        except IndexError:  # there is no socket available: lets create one
-            socket = context.socket(zmq.REQ)
-            socket.connect(self.socket_path)
-        try:
-            yield socket
-        finally:
-            if not socket.closed:
-                self._sockets.append((socket, time.time()))
+        if not pt_socket:
```

Comment:
Maybe you can create a function to avoid the duplicated code into the « if » « else ».